src/frontend/core.ts

```diff
@@ -405,6 +412,35 @@ export function pad(x: TracerValue, width: number | Pair | Pair[]) {
   return bind1(Primitive.Pad, [x], { width });
 }
 
+/**
+ * Solve a triangular linear system.
+ *
+ * Solves `a @ x = b` (if leftSide=true) or `x @ a = b` (if leftSide=false)
+ * where `a` is a triangular matrix.
+ */
+export function triangularSolve(
+  a: TracerValue,
+  b: TracerValue,
+  {
+    leftSide = true,
+    lower = true,
+    transposeA = false,
+    unitDiagonal = false,
+  }: {
+    leftSide?: boolean;
+    lower?: boolean;
+    transposeA?: boolean;
+    unitDiagonal?: boolean;
+  } = {},
+) {
```

Comment:
I think we should try to minimize the number of parameters passed to primitives. In this case: `transposeA`, `leftSide` can both be implemented with movement ops and this will reduce the amount of code we need to write/maintain